packages/core/src/scheduler.ts

```diff
@@ -0,0 +1,113 @@
+import { AnyActorRef, EventObject } from '.';
+import { AnyActorSystem } from './system';
+
+export interface ScheduledEvent {
+  id: string;
+  event: EventObject;
+  startedAt: number; // timestamp
+  delay: number;
+  source: AnyActorRef;
+  target: AnyActorRef;
+}
+
+export interface Clock {
+  setTimeout(fn: (...args: any[]) => void, timeout: number): any;
+  clearTimeout(id: any): void;
+}
+
+export interface Scheduler {
+  events: { [id: string]: ScheduledEvent };
+  schedule(
+    source: AnyActorRef,
+    data: {
+      id: string | undefined;
+      event: EventObject;
+      delay: number;
+      to?: AnyActorRef;
+    }
+  ): void;
+  cancel(id: string): void;
+  cancelAll(actorRef: AnyActorRef): void;
+  getPersistedSnapshot(): Record<string, ScheduledEvent>;
+  start(): void;
+}
+
+const defaultClock: Clock = {
+  setTimeout: (fn, ms) => {
+    return setTimeout(fn, ms);
+  },
+  clearTimeout: (id) => {
+    return clearTimeout(id);
+  }
+};
+
+export function createScheduler(
+  clock: Clock,
+  system: AnyActorSystem,
+  snapshot: unknown
+): Scheduler {
+  const timerMap: { [id: string]: number } = {};
+
+  const scheduler: Scheduler = {
+    events: (snapshot as any) ?? {},
+    schedule: (
+      source: AnyActorRef,
+      data: {
+        id: string | undefined;
+        event: EventObject;
+        delay: number;
+        to: AnyActorRef;
+      }
+    ) => {
+      // TODO: `id` has to be separated from `data.id` completely
+      // `data.id` is supposed to be unique within an actor, `id` is supposed to be unique globally
+      const id = data.id ?? Math.random().toString(36).slice(2);
+      const scheduledEvent: ScheduledEvent = {
+        ...data,
+        id,
+        source,
+        target: data.to || source,
+        startedAt: Date.now()
+      };
+      scheduler.events[id] = scheduledEvent;
+
+      const timeout = clock.setTimeout(() => {
+        const target = data.to || source;
+        // TODO: explain this hack, it should also happen sooner, not within this timeout
```

Comment:
this likely is fixable (but I didn't check it) by deferring `scheudle`+`cancel` in our machine actions

Reply:
done